cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java

```diff
@@ -0,0 +1,212 @@
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.lib.AbstractCloseableIterator;
+import co.cask.cdap.api.dataset.lib.AbstractDataset;
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.api.dataset.module.EmbeddedDataset;
+import co.cask.cdap.api.dataset.table.Put;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.internal.app.runtime.schedule.trigger.TriggerJsonCodec;
+import co.cask.cdap.internal.schedule.trigger.Trigger;
+import co.cask.cdap.proto.Notification;
+import co.cask.cdap.proto.id.ScheduleId;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Dataset that stores {@link Job}s, which correspond to schedules that have been triggered, but not yet executed.
+ *
+ * Row Key is in the following formats:
+ *   For Jobs:
+ *     'J':<partition_id>:<scheduleId>:<timestamp>
+ *     The <partition_id> is a hash based upon the scheduleId
+ *
+ *   For TMS MessageId:
+ *     'M':<topic>
+ */
+public class JobQueueDataset extends AbstractDataset implements JobQueue {
+
+  static final String EMBEDDED_TABLE_NAME = "t"; // table
+  private static final Gson GSON =
+    new GsonBuilder().registerTypeAdapter(Trigger.class, new TriggerJsonCodec()).create();
+
+  // For now, simply serialize the entire Job into one column
+  private static final byte[] COL = new byte[] {'C'};
+  private static final byte[] JOB_ROW_PREFIX = new byte[] {'J'};
+  private static final byte[] ROW_KEY_SEPARATOR = new byte[] {':'};
+  private static final byte[] MESSAGE_ID_ROW_PREFIX = new byte[] {'M'};
+
+  private static final int NUM_PARTITIONS = 16;
+
+  private final Table table;
+
+  JobQueueDataset(String instanceName, @EmbeddedDataset(EMBEDDED_TABLE_NAME) Table table) {
+    super(instanceName, table);
+    this.table = table;
+  }
+
+  @Override
+  public CloseableIterator<Job> getJobsForSchedule(ScheduleId scheduleId) {
+    byte[] keyPrefix = getRowKeyPrefix(scheduleId);
+    return createCloseableIterator(table.scan(keyPrefix, Bytes.stopKeyForPrefix(keyPrefix)));
+  }
+
+  @Override
+  public Job getJob(JobKey jobKey) {
+    Row row = table.get(getRowKey(jobKey.getScheduleId(), jobKey.getCreationTime()));
+    if (row.isEmpty()) {
+      return null;
+    }
+    return fromRow(row);
+  }
+
+  @Override
+  public void put(Job job) {
+    table.put(toPut(job));
+  }
+
+  @Override
+  public void updateState(Job job, Job.State state) {
+    // assert that the job state transition is valid
+    job.getState().check(state);
+    put(new SimpleJob(job.getSchedule(), job.getCreationTime(), job.getNotifications(), state));
```

Comment:
Right, I didn't mean users of CDAP, but just whoever is maintaining/modifying any interaction with the Job class.
To your second point, I'm not sure how we can handle other types of Jobs in the updateState, without knowing what the other (future) implementations will be.